proposals/0000-warning-pragmas-with-categories.rst

```diff
@@ -0,0 +1,290 @@
+users.
+
+Instead, this ghc-proposal extends ``WARNING`` pragmas so they can be annotated
+with a "warning category".  Users can then enable or disable these
+categories individually.  For example, it would allow the core-libraries
+proposal to introduce::
+
+    {-# WARNING [partial] head "This is a partial function, it throws an error on empty lists." #-}
+    {-# WARNING [partial] tail "This is a partial function, it throws an error on empty lists." #-}
+
+Such warnings are enabled by default, so the appropriate warning would be
+displayed at any use of ``head`` or ``tail``.  However, users may control this
+behaviour to their liking, prefixing the warning category name with an ``x-`` to
+distinguish it from the built-in warning flags:
+
+* Users who prefer to use the functions unimpeded may specify
+  ``-Wno-x-partial`` to suppress all warnings annotated with the
+  ``partial`` warning category, while still seeing other warnings.
+
+* Users who do not wish to use the functions at all may specify
+  ``-Werror=x-partial`` so that any use of the functions will be reported
+  as an error, while other warnings are unaffected.
+
+
+Proposed Change Specification
+-----------------------------
+
+The `WARNING pragma
+<https://downloads.haskell.org/ghc/9.4.1/docs/users_guide/exts/pragmas.html#warning-deprecated-pragma>`_
+may be immediately followed by a single warning category in
+square brackets. A warning category is an identifier beginning with a letter
+and followed by valid identifier characters or dashes.
+
+More formally, the grammar of declarations is extended as follows:
+
+============  =  =====================================================================
+*decl*        →  ``{-# WARNING`` [*categories*] [*things*] *strings* ``#-}``
+
+*categories*  →  ``[`` *category* ``]``
+*category*    →  (*small* | *large*) { *small* | *large* | *digit* | ``'`` | ``-`` }
+
+*things*      →  *thing1*, ..., *thingN*
+*thing*       →  *varid* | *conid*
+*strings*     →  *string* | ``[`` *string1*, ..., *stringN* ``]``
+============  =  =====================================================================
+
+The category can be omitted entirely, so this subsumes the existing
+syntax for ``WARNING`` pragmas.  That is, [*categories*]
+means that the presence of the *categories* non-terminal is optional; if present
+it is a single category in square brackets (hence the use of typewriter font ``[``
+... ``]`` in the definition of the *categories* production).
+Similarly the list of *things* is optional as it may be omitted (for a ``WARNING`` on a module header).
+
+The *category* non-terminal subsumes both *varid* and *conid*, so it is
+possible to use the name of the thing to which a warning is being attached as
+the category, provided it is not an operator.  The dash character (``-``) is permitted as a character
+in addition to identifier characters, since dashes are frequently used in
+warning names.
+
```

Comment:
```suggestion
The *category* non-terminal subsumes both *varid* and *conid*, so it is
possible to use the name of the thing to which a warning is being attached as
the category, provided it is not an operator.  The dash character (``-``) is permitted as a character
in addition to identifier characters, since dashes are frequently used in
warning names.

The compiler only allows a *category* that starts with `x-` or is `deprecations`.
(This restriction is not part of the syntax change because we may expect later
extensions to that list.)


```